csrc/fusion_profiler.cpp

```diff
@@ -0,0 +1,737 @@
+#include <iomanip>
+
+namespace nvfuser {
+
+namespace {
+
+// Copying some code from the CUPTI samples/common code
+// CUPTI buffer size 4.0 KB
+#define BUF_SIZE (1 * 4 * 1024)
+// 8-byte alignment for the buffers
+#define ALIGN_SIZE (8)
+#define ALIGN_BUFFER(buffer, align)                                 \
+  (((uintptr_t)(buffer) & ((align)-1))                              \
+       ? ((buffer) + (align) - ((uintptr_t)(buffer) & ((align)-1))) \
+       : (buffer))
+
+const char* get_demangled_name(const char* pName) {
+  if (pName == nullptr) {
+    return "<null>";
+  }
+  int status = 0;
+  return abi::__cxa_demangle(pName, nullptr, nullptr, &status);
+}
+
+void record_cupti_activity(CUpti_Activity* pRecord, FILE* pFileHandle) {
+  CUpti_ActivityKind activityKind = pRecord->kind;
+
+  switch (activityKind) {
+    case CUPTI_ACTIVITY_KIND_CONCURRENT_KERNEL: {
+      CUpti_ActivityKernel8* pKARecord = (CUpti_ActivityKernel8*)pRecord;
+
+      KernelProfile prof;
+      prof.name.assign(get_demangled_name(pKARecord->name));
+      size_t start = prof.name.find("kernel");
+      size_t end = prof.name.find('(');
+      prof.name = prof.name.substr(start, end - start);
+      prof.device = (int)pKARecord->deviceId;
+      prof.stream = pKARecord->streamId;
+      prof.correlation_id = pKARecord->correlationId;
+      constexpr double ms_convert = 1.0 / 1000000.0;
+      prof.time_ms =
+          static_cast<double>(pKARecord->end - pKARecord->start) * ms_convert;
+      prof.grid = {pKARecord->gridX, pKARecord->gridY, pKARecord->gridZ};
+      prof.block = {pKARecord->blockX, pKARecord->blockY, pKARecord->blockZ};
+      prof.cluster = {
+          pKARecord->clusterX, pKARecord->clusterY, pKARecord->clusterZ};
+      prof.dynamic_shared_mem = pKARecord->dynamicSharedMemory;
+      prof.static_shared_mem = pKARecord->staticSharedMemory;
+      prof.registers = pKARecord->registersPerThread;
+
+      FusionProfiler::get()->recordAsyncKernelActivity(std::move(prof));
+
+      break;
+    }
+    case CUPTI_ACTIVITY_KIND_EXTERNAL_CORRELATION: {
+      CUpti_ActivityExternalCorrelation* pExternalCorrelationRecord =
+          (CUpti_ActivityExternalCorrelation*)pRecord;
+      FusionProfiler::get()->recordAsyncCorrIdActivity(
+          pExternalCorrelationRecord->externalId,
+          pExternalCorrelationRecord->correlationId);
+      break;
+    }
+    case CUPTI_ACTIVITY_KIND_DRIVER:
+      break;
+    default:
+      fprintf(pFileHandle, "  <unknown>\n");
+      break;
+  }
+}
+
+void record_cupti_activity_buffer(
+    uint8_t* pBuffer,
+    size_t validBytes,
+    FILE* pFileHandle,
+    void* pUserData) {
+  CUpti_Activity* pRecord = nullptr;
+  CUptiResult status = CUPTI_SUCCESS;
+
+  do {
+    status = cuptiActivityGetNextRecord(pBuffer, validBytes, &pRecord);
+    if (status == CUPTI_SUCCESS) {
+      record_cupti_activity(pRecord, stdout);
+    } else if (status == CUPTI_ERROR_MAX_LIMIT_REACHED) {
+      break;
```

Comment:
Dumb question: What happens when max limit is reached? should we throw a warning here?

Reply:
It walks the record in order to reach the limit.  It always reaches this point.

Reply:
I added some more security to this loop so it cannot go into an infinite loop and I commented the `status == CUPTI_ERROR_MAX_LIMIT_REACHED` case.